components/backup_text/src/mask.rs

```
@@ -0,0 +1,195 @@
+    } else if let Ok(_) = u16::try_from(from) {
+        to % (u16::MAX as u64 + 1)
+    } else if let U24_MIN..=U24_MAX = from {
+        to % (U24_MAX as u64 + 1)
+    } else if let Ok(_) = u32::try_from(from) {
+        to % (u32::MAX as u64 + 1)
+    } else {
+        to
+    }
+}
+
+#[inline]
+fn mask_f64(f: f64) -> f64 {
+    f64::from_le_bytes(hash_bytes(&mut f.to_le_bytes(), 8).try_into().unwrap())
+}
+
+fn mask_string(bytes: &[u8]) -> Vec<u8> {
+    let size = bytes.len();
+    let sum = hash_bytes(bytes, size / 2);
+    let mut hex = hex::encode(sum);
+    hex.push_str(&"*".repeat(size - hex.len()));
+    hex.into_bytes()
+}
+
+fn mask_duration(dur: Duration) -> Duration {
+    let secs = dur.to_secs() / 3600 * 3600;
+    let fsp = dur.fsp() as i8;
+    Duration::from_secs(secs, fsp).unwrap()
+}
+
+fn mask_time(time: DateTime) -> DateTime {
+    let ctx = &mut eval_context();
+
+    let year = time.year();
+    let month = time.month();
+    let day = time.day();
+    let fsp = time.fsp() as i8;
+    let time_type = time.get_time_type();
+
+    DateTime::parse(
+        ctx,
+        &format!("{}-{}-{}", year, month, day),
+        time_type,
+        fsp,
+        true,
+    )
+    .unwrap()
+}
+
+pub fn workload_sim_mask(mut datum: Datum) -> MaskResult {
+    match &mut datum {
+        Datum::Null => {}
+        Datum::Min => {}
+        Datum::Max => {}
+        Datum::I64(i) => *i = mask_i64(*i),
+        Datum::U64(u) => *u = mask_u64(*u),
+        Datum::F64(f) => *f = mask_f64(*f),
+        Datum::Bytes(bytes) => *bytes = mask_string(bytes).to_vec(),
+        Datum::Enum(e) => {
+            let masked_name = mask_string(e.name()).to_vec();
+            *e = Enum::new(masked_name, e.value())
+        }
+        Datum::Dur(dur) => *dur = mask_duration(*dur),
+        Datum::Time(time) => *time = mask_time(*time),
+
+        Datum::Dec(_) | Datum::Json(_) | Datum::Set(_) => {
+            return Err(datum); // todo: not supported yet
```

We don't need to return an error that was never handled, how about just changing the function to `fn workload_sim_mask(mut datum: Datum) -> Datum`

---

OK. My first thought is that we have to tell the outside world about the failure of masking. However it seems nonsense to fail to mask a `Datum` in production, let's just always return with a success.

---

It seems fine to me to return the failure of masking, but in this way, we need to handle that failure (panic, logging or return an error to the upper caller), not just use the non-mask datum and pretend no error return. Also the current possible error of `Dec`, `Json`, `Set` is more like `not support yet` error, keeping that error inside the masking function is enough.